extension/extension.proto

```diff
@@ -0,0 +1,177 @@
+      returns (UninstallExtensionResponse);
+}
+
+message FinalizeExtensionsRequest{
+}
+message FinalizeExtensionsResponse{
+}
+
+message SigVerificationResponse{
+}
+
+message SigVerificationRequest{
+  enum TrustSource {
+    // UNSPECIFIED should be treated as an error.
+    UNSPECIFIED = 0;
+    // ANY = disable signature verification. Allows any extension
+    // (even one with missing/invalid signature) to be installed
+    ANY = 1;
+    // MANUFACTURER = trust only extensions which have been signed
+    // by the device manufacturer to be installed.
+    MANUFACTURER = 2;
+    // SSL_PROFILE = trust extensions which are valid according to
+    // the SSL profile specified by the ssl_profile_id field.
+    // creating and populating SSL profiles is out of the scope of this
+    // service and should be done through some other means (e.g. gNSI Certz).
+    SSL_PROFILE = 3;
+  }
+  TrustSource source =  1;
+
+  // Optional ssl_profile_id field, which should be populated only
+  // if the source of trust is an SSL profile.
+  // The server should ignore this field if source is not SSL_PROFILE
+  optional string ssl_profile_id = 2;
+}
+
+message UninstallExtensionRequest{
+  // name of the extension to be uninstalled
+  string name = 1;
+
+  // a boolean indicating whether we should force uninstalling this extension.
+  // if set to true, then ignore conflicts and dependencies when uninstalling.
+  bool force = 2;
+}
+message UninstallExtensionResponse{
+}
+
+message InstallExtensionResponse{
+  enum Operation{
+    UNSPECIFIED = 0;
+    TRANSFER_COMPLETE = 1;
+    INSTALL_COMPLETE = 2;
+  }
+  Operation acknowledgement = 3;
+}
+
+message InstallExtensionRequest{
+  oneof request{
+    TransferRequest transfer_request = 1;
+    InstallRequest install_request = 2;
+  }
+}
+
+// InstallRequest is a request for installing an extension onto a system.
+// The extension is applied to the running configuration.
+message InstallRequest{
+  // extension_name is the name of the extension to install.
+  // This is required because InstallExtension can be called
+  // for an extension which is already present on the system
+  string extension_name = 1;
+
+  // if this is true, then skip signature verification
+  // for this particular extension (i.e. allow missing or invalid signatures)
+  bool ignore_signature_verification = 2;
+
+  // if this is true, then the system will also
+  // ensure that this extension is loaded on boot/reboot
+  bool load_after_boot = 3;
+
+  // if this is true, then the system will ignore
+  // any errors which could arise from missing dependencies
+  bool ignore_dependency_checks = 4;
+}
+
+message TransferRequest{
+  oneof request {
+    // name of the extension  (including the file extension)
+    string extension_name = 1;
+    // raw byte contents to be streamed into the extension file.
+    bytes contents = 2;
+    // hash of the file.
+    types.HashType hash = 3;
+  }
+}
```

Comment:
I wonder if we want to have multiple file transfer ops spread across different services.

System.Package - https://gnxi.srlinux.dev/gnoi/system#SetPackageRequest looks to be a close match

Reply:
>I wonder if we want to have multiple file transfer ops spread across different services.

I am ok with having multiple file transfer ops across services, mainly because:
1. [gNSI Authz](https://github.com/openconfig/gnsi/tree/a80d61cd116210b0dfcac021ecd74906f3bbf4fa/authz) is service/RPC based. If we delegate that all file transfer operations must be done by some common RPC (e.g. `gNOI File/Put` RPC), then this means that for a system running gNSI Authz, all clients must also be authorized to run the `gNOI File/Put` RPC, which would enable them to put any file on the system that they want.
I think its better to build the file-transfer operations into the RPC itself, so that the a client created for the purpose of "transferring extensions onto the system" is only given the ability to do that, rather than general access to the whole filesystem. 

2. (smaller point) for ease of use, i like that a single client (one for only the `Extension` service) can be used to interact with a service, rather than creating multiple clients (one for `File` and another for `Extension`)

>System.Package - https://gnxi.srlinux.dev/gnoi/system#SetPackageRequest looks to be a close match

I mention this in a separate comment here: https://github.com/openconfig/gnoi/pull/160/#discussion_r1472784052